src/buffer.rs

```diff
@@ -0,0 +1,134 @@
+
+/// A buffer manages a fixed amount of data.
+///
+/// The data are produced one at a time or consumed all at once.
+pub trait Buffer<T> {
+    /// Add the given element to the buffer if possible.
+    fn add(&mut self, element: T) -> Result<(), BufferFullError<T>>;
+
+    /// Copy and return all elements of the buffer.
+    ///
+    /// When this operation completes, the buffer state will reset so that the buffer
+    /// has no elements. In other words, the buffer elements are completely consumed
+    /// and ownership transferred.
+    ///
+    /// Note that the underlying memory is *not* zeroed, so sensitive information should
+    /// not be stored in this buffer.
+    fn consume(&mut self) -> Option<Vec<T>>;
+}
+
+const BUFFER_SIZE: usize = 512;
+
+/// Buffer whose capacity is determined at compile time.
+pub struct StaticBuffer<T>
+where
+    T: Copy + Default + Sized,
+{
+    elements: [T; BUFFER_SIZE],
+    length: usize,
+}
+
+impl<T> StaticBuffer<T>
+where
+    T: Copy + Default,
+{
+    /// Constructs a new buffer.
+    pub fn new() -> StaticBuffer<T> {
+        StaticBuffer {
+            elements: [Default::default(); BUFFER_SIZE],
+            length: 0,
+        }
+    }
+}
+
+impl<T> Buffer<T> for StaticBuffer<T>
+where
+    T: Copy + Default,
+{
+    fn add(&mut self, element: T) -> Result<(), BufferFullError<T>> {
+        if self.length == BUFFER_SIZE {
+            return Err(BufferFullError(element));
+        }
+        self.elements[self.length] = element;
+        self.length += 1;
+        Ok(())
+    }
+
+    fn consume(&mut self) -> Option<Vec<T>> {
+        if self.length == 0 {
+            return None;
+        }
+        let mut data = vec![Default::default(); self.length];
+        data.copy_from_slice(&self.elements[..self.length]);
+        self.length = 0;
+        Some(data)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn static_buffer_new() {
+        let buffer: StaticBuffer<i32> = StaticBuffer::new();
+        assert_eq!(buffer.length, 0);
+    }
+
+    #[test]
+    fn static_buffer_add_empty() {
+        let mut buffer: StaticBuffer<usize> = StaticBuffer::new();
+        let expected = 123;
+        buffer.add(expected).unwrap();
+        assert_eq!(buffer.length, 1);
+        assert_eq!(buffer.elements[0], expected);
+    }
+
+    #[test]
+    fn static_buffer_add_full() {
+        let mut buffer: StaticBuffer<usize> = StaticBuffer::new();
+
+        for value in 0..BUFFER_SIZE {
+            buffer.add(value).unwrap();
+        }
+        assert_eq!(buffer.length, BUFFER_SIZE);
+
+        let actual = buffer.elements.to_vec();
+        let expected: Vec<usize> = (0..BUFFER_SIZE).collect();
+        assert_eq!(actual, expected);
+
+        let expected = BUFFER_SIZE;
+        if let Err(BufferFullError(actual)) = buffer.add(expected) {
+            assert_eq!(actual, expected);
+        } else {
+            panic!("should have returned error");
+        }
+    }
```

Comment:
This is testing two separate functionalities of the `add` function (add to buffer and error if full). Should these be separate tests? (I'm honestly asking, I don't know)

Reply:
The add-to-buffer functionality is isolated in a separate test, so I think this is okay. You can view this test as actually testing: what happens if I repeatedly add() to the queue?